File: leetcode/problems/language-specific/go/62.unique-paths.go
/*
 * @lc app=leetcode id=62 lang=golang
 *
 * [62] Unique Paths
 */

// @lc code=start

// * Bottom-Up Iterative Solution | O(m^n) Time | O(n) Space
// * m -> Number of rows | n -> Number of columns

func uniquePaths(m int, n int) int {
	if m == 0 || n == 0 {
		return 0
	}

	dp := make([]int, n)
	for i := 0; i < n; i++ {
		dp[i] = 1
	}

	for i := 1; i < m; i++ {
		for j := 1; j < n; j++ {
			dp[j] += dp[j-1]
		}
	}

	return dp[n-1]
}

// @lc code=end

File: leetcode/problems/language-specific/python/62.unique-paths.py
#
# @lc app=leetcode id=62 lang=python3
#
# [62] Unique Paths
#

# @lc code=start

# * Iterative Bottom-Up Solution | O(mn) Time | O(n) Space
# * m -> The given input | n -> The given input


class Solution:
    def uniquePaths(self, m: int, n: int) -> int:
        # * Create a 1-D DP array with the size as n and the values as 1 since
        # * the first row and first col will have only 1 possible pathway.
        dp = [1] * n
        # * Start from the first row and go through all the cols except for the first one.
        for _ in range(1, m):
            for col in range(1, n):
                dp[col] += dp[col - 1]

        return dp[n - 1]


# @lc code=end
